File: rover.rb
#!/usr/bin/env ruby

# Init with field size and location/direction.
# Accept commands to M(ove) and rotate L(eft) and R(ight).
# Ignore commands to move off the field.

class Rover
  attr_accessor :w, :h, :x, :y, :d, :field
  DIR = %w[N E S W]
  CUR = %W[▲ ▶ ▼ ◀]

  def initialize(w, h, x, y, c)
    self.w = w
    self.h = h
    self.x = x
    self.y = y
    self.d = DIR.index(c) || 0
    self.field = Array.new(h) { Array.new(w) { "." } }
    plot
    puts " #{x} #{y} (#{DIR[d]})"
  end

  def move(cmd)
    cmd.each_char do |c|
      if c == "M"
        if d == 0 && y < h - 1 # N
          self.y += 1
        elsif d == 1 && x < w - 1 # E
          self.x += 1
        elsif d == 2 && y > 0 # S
          self.y -= 1
        elsif d == 3 && x > 0 # W
          self.x -= 1
        end
      elsif c == "L"
        self.d = d == 0 ? 3 : d - 1
      elsif c == "R"
        self.d = d == 3 ? 0 : d + 1
      end
      plot
    end
    puts " #{cmd}\n #{x} #{y} (#{DIR[d]})"
    show
  end

  def plot
    field[y][x] = CUR[d]
  end

  def show
    print "\n"
    (h - 1).downto(0) do |r|
      print r > 9 ? r.to_s : " #{r}"
      0.upto(w - 1) do |c|
        print " #{field[r][c]}"
      end
      print "\n"
    end
    print "  "
    0.upto(field.first.size - 1) { |n| print " #{n % 10}"}
    print "\n\n"
  end
end

rover = Rover.new(5, 5, 1, 1, "N")
rover.move("MLMMMMMRMRMMMMMMMRMM")

__END__

 1 1 (N)
 MLMMMMMRMRMMMMMMMRMM
 4 1 (S)

 4 . . . . .
 3 ▶ ▶ ▶ ▶ ▼
 2 ▲ ◀ . . ▼
 1 . ▲ . . ▼
 0 . . . . .
   0 1 2 3 4

File: rover.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Init with field size and location/direction.
# Accept commands to M(ove) and rotate L(eft) and R(ight).
# Ignore commands to move off the field.

DIR = ["N", "E", "S", "W"]
CUR = ["↑", "→", "↓", "←"]


class Rover:

    def __init__(self, w, h, x, y, c):
        self.w = w
        self.h = h
        self.x = x
        self.y = y
        self.d = DIR.index(c)
        self.field = []
        for r in range(h):
            self.field.append(["."] * w)
        self.plot()
        print(" %s %s (%s)" % (x, y, DIR[self.d]))

    def move(self, cmd):
        for c in cmd:
            if c == "M":
                if self.d == 0 and self.y < self.h - 1:  # N
                    self.y += 1
                elif self.d == 1 and self.x < self.w - 1:  # E
                    self.x += 1
                elif self.d == 2 and self.y > 0:  # S
                    self.y -= 1
                elif self.d == 3 and self.x > 0:  # W
                    self.x -= 1
            elif c == "L":
                self.d = 3 if self.d == 0 else self.d - 1
            elif c == "R":
                self.d = 0 if self.d == 3 else self.d + 1
            self.plot()
        print(" %s\n %s %s (%s)" % (cmd, self.x, self.y, DIR[self.d]))
        self.show()

    def plot(self):
        self.field[self.y][self.x] = CUR[self.d]

    def show(self):
        print
        for r in reversed(range(self.h)):
            print(r if r > 9 else " %s" % r),
            for c in self.field[r]:
                print(c),
            print
        print("  "),
        for n in range(self.w):
            print(n % 10),
        print


rover = Rover(5, 5, 1, 1, "N")
rover.move("MLMMMMMRMRMMMMMMMRMM")

"""
 1 1 (N)
 MLMMMMMRMRMMMMMMMRMM
 4 1 (S)

 4 . . . . .
 3 ▶ ▶ ▶ ▶ ▼
 2 ▲ ◀ . . ▼
 1 . ▲ . . ▼
 0 . . . . .
   0 1 2 3 4
"""
